pmd-cli/src/main/java/net/sourceforge/pmd/cli/commands/internal/AbstractAnalysisPmdSubcommand.java

```diff
@@ -34,6 +33,14 @@ public abstract class AbstractAnalysisPmdSubcommand<C extends AbstractConfigurat
             arity = "1..*", split = ",")
     protected List<Path> inputPaths;
 
+    @Parameters(arity = "*", description = "Path to a source file, or directory containing source files to analyze. "
+            + "Equivalent to using --dir.")
+    // Note: we can't directly use inputPaths here, because "--dir" might be specified later in the
+    // command line. And PicoCli will then initialize the field inputPaths with a fresh empty List, replacing
+    // anything what was added before.So we need to combine the two lists in #validate()
+    private List<Path> positionalInputPaths;
```

Comment:
I'm unsure I follow… positional parameters are always at the end. Any `--dir` occurrences must be previous…

Reply:
apparently, every unmatched parameter is taken as positional, which means these positional parameters can appear at any location...

E.g. `--minimum-tokens 34 --dir SRC_DIR --ignore-identifiers false --debug` was executed before my change as follows:
1. dir: sets "inputPaths" to contain SRC_DIR
2. false: resets the "inputPaths" to only contain false

I don't remember the exact details, but it's implementation specific to PicoCli (I debugged a lot). PicoCli at some point re-initializes the fields before calling the setter or so... the net result was, that SRC_DIR got removed, because PicoCli reset the field to an empty List again before adding/processing the positional parameters...

Reply:
Have you tried using a setter for both `--dir` and positional parameters? Reflection to override the value should not be used if a setter is present, and the old `public void setInputPaths(final List<Path> inputPaths)` handled the null case… so maybe having the same for `--dir` would work, without relying on calling `validate` for something that is not a validation.

Reply:
I'll give it a try - would be good, if we don't need to depend on `validate` being called....

Reply:
This seems to be working - at lest the unit tests are still good.

I'm going to merge this PR now.